In [24]:
import os
import numpy as np 
import pandas as pd
from sklearn.model_selection import cross_val_predict
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
from sklearn import svm
import matplotlib.pyplot as plt

In [25]:
def model(train):
    model=LogisticRegression(random_state=1,class_weight="balanced")
    #model = svm.LinearSVC(class_weight="balanced", random_state=1)
    predictors=train.columns.tolist()
    excluded_features=["foreclosure_status", "id", "origination_date", "origination_month", "origination_year"]
    predictors = [p for p in predictors if p not in excluded_features]

    predictions = cross_val_predict(model,train[predictors],train["foreclosure_status"],cv=3)

    return predictions

In [26]:
def compute_error(target,predictions):
    return metrics.accuracy_score(target,predictions)

In [27]:
def compute_false_negatives(target,predictions):
    false_negatives=pd.DataFrame({"target":target,"predictions":predictions})
    neg_rate=false_negatives[(false_negatives["target"]==1) & (false_negatives["predictions"]==0)].shape[0]/(false_negatives[(false_negatives["target"]==1)].shape[0]+1)
    return neg_rate

In [28]:
def compute_false_positive(target,predictions):
    false_positives=pd.DataFrame({"target":target,"predictions":predictions})
    pos_rate=false_positives[(false_positives["target"]==0) & (false_positives["predictions"]==1)].shape[0]/(false_positives[(false_positives["target"]==0)].shape[0]+1)
    return pos_rate

In [29]:
def read():
    PROCESSED_DIR = "/home/seb/Documents/prepayment/processed/"
    train=pd.read_csv(os.path.join(PROCESSED_DIR,"train.csv"))
    return train

In [31]:
train=read()
train=train.drop(["origination_date"],1)
predictions=model(train)
model_error=compute_error(train["foreclosure_status"],predictions)
FN=compute_false_negatives(train["foreclosure_status"],predictions)
FP=compute_false_positive(train["foreclosure_status"],predictions)
print("Accuracy of the model:{}".format(model_error))
print("False Negatives:{}".format(FN))
print("False Positive:{}".format(FP))

#cm = confusion_matrix(train["foreclosure_status"],predictions)
print(cm)


/home/seb/anaconda3/envs/ml/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/home/seb/anaconda3/envs/ml/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/home/seb/anaconda3/envs/ml/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Accuracy of the model:0.7081416579062509
False Negatives:0.2657210401891253
False Positive:0.2918874671149677


NameError: name 'cm' is not defined

Accuracy of the model:0.7081416579062509

False Negatives:0.2657210401891253

False Positive:0.2918874671149677